File: Pharmaceutical_Company_Example/03_Behavioral/PHP/Mediator.php
<?php

// Mediator Interface
interface Mediator
{
    public function notify(string $sender, string $event): void;
}

// Concrete Mediator
class RegulatoryMediator implements Mediator
{
    private ?ClinicalTrials $clinicalTrials = null;
    private ?QualityAssurance $qualityAssurance = null;
    private ?RegulatoryAffairs $regulatoryAffairs = null;
    private ?LegalDepartment $legal = null;

    public function notify(string $sender, string $event): void
    {
        switch ($event) {
            case "Clinical Trial Completed":
                echo "Mediator: Clinical Trial completed. Notifying Quality Assurance.\n";
                if ($this->qualityAssurance) {
                    $this->qualityAssurance->checkQuality();
                }
                break;

            case "Quality Check Completed":
                echo "Mediator: Quality Check completed. Notifying Regulatory Affairs.\n";
                if ($this->regulatoryAffairs) {
                    $this->regulatoryAffairs->prepareDocuments();
                }
                break;

            case "Documents Prepared":
                echo "Mediator: Documents prepared. Notifying Legal Department.\n";
                if ($this->legal) {
                    $this->legal->reviewDocuments();
                }
                break;

            case "Legal Review Completed":
                echo "Mediator: Legal Review completed. Drug is ready for final approval.\n";
                break;
        }
    }

    // Setters for departments
    public function setClinicalTrials(ClinicalTrials $clinicalTrials): void
    {
        $this->clinicalTrials = $clinicalTrials;
    }

    public function setQualityAssurance(QualityAssurance $qualityAssurance): void
    {
        $this->qualityAssurance = $qualityAssurance;
    }

    public function setRegulatoryAffairs(RegulatoryAffairs $regulatoryAffairs): void
    {
        $this->regulatoryAffairs = $regulatoryAffairs;
    }

    public function setLegal(LegalDepartment $legal): void
    {
        $this->legal = $legal;
    }
}

// Department Interfaces
abstract class Department
{
    protected Mediator $mediator;

    public function __construct(Mediator $mediator)
    {
        $this->mediator = $mediator;
    }

    abstract public function completeTask(): void;
}

// Concrete Department: Clinical Trials
class ClinicalTrials extends Department
{
    public function runTrials(): void
    {
        echo "Clinical Trials: Running clinical trials...\n";
        $this->mediator->notify("ClinicalTrials", "Clinical Trial Completed");
    }

    public function completeTask(): void
    {
        // Task completion logic
    }
}

// Concrete Department: Quality Assurance
class QualityAssurance extends Department
{
    public function checkQuality(): void
    {
        echo "Quality Assurance: Checking product quality...\n";
        $this->mediator->notify("QualityAssurance", "Quality Check Completed");
    }

    public function completeTask(): void
    {
        // Task completion logic
    }
}

// Concrete Department: Regulatory Affairs
class RegulatoryAffairs extends Department
{
    public function prepareDocuments(): void
    {
        echo "Regulatory Affairs: Preparing regulatory documents...\n";
        $this->mediator->notify("RegulatoryAffairs", "Documents Prepared");
    }

    public function completeTask(): void
    {
        // Task completion logic
    }
}

// Concrete Department: Legal Department
class LegalDepartment extends Department
{
    public function reviewDocuments(): void
    {
        echo "Legal Department: Reviewing regulatory documents...\n";
        $this->mediator->notify("LegalDepartment", "Legal Review Completed");
    }

    public function completeTask(): void
    {
        // Task completion logic
    }
}

// Create the Mediator
$mediator = new RegulatoryMediator();

// Create Departments
$clinicalTrials = new ClinicalTrials($mediator);
$qualityAssurance = new QualityAssurance($mediator);
$regulatoryAffairs = new RegulatoryAffairs($mediator);
$legal = new LegalDepartment($mediator);

// Set Departments in the Mediator
$mediator->setClinicalTrials($clinicalTrials);
$mediator->setQualityAssurance($qualityAssurance);
$mediator->setRegulatoryAffairs($regulatoryAffairs);
$mediator->setLegal($legal);

// Run process
$clinicalTrials->runTrials();

?>

File: Pharmaceutical_Company_Example/03_Behavioral/C++/Mediator.cpp
#include <iostream>
#include <memory>
#include <string>

using namespace std;

// Mediator Interface
class Mediator {
public:
  virtual void notify(const string& sender, const string& event) = 0;
  virtual ~Mediator() = default;
};

// Concrete Mediator
class RegulatoryMediator : public Mediator {
private:
  class ClinicalTrials* clinical_trials;
  class QualityAssurance* quality_assurance;
  class RegulatoryAffairs* regulatory_affairs;
  class LegalDepartment* legal;
  
public:
  void setClinicalTrials(ClinicalTrials* ct) { clinical_trials = ct; }
  void setQualityAssurance(QualityAssurance* qa) { quality_assurance = qa; }
  void setRegulatoryAffairs(RegulatoryAffairs* ra) { regulatory_affairs = ra; }
  void setLegalDepartment(LegalDepartment* ld) { legal = ld; }
  
  void notify(const string& sender, const string& event) override;
};

// Department Interfaces
class Department {
protected:
  Mediator* mediator;
public:
  Department(Mediator* mediator) : mediator(mediator) {}
  virtual void completeTask() = 0;
  virtual ~Department() = default;
};

// Concrete Department: Clinical Trials
class ClinicalTrials : public Department {
public:
  ClinicalTrials(Mediator* mediator) : Department(mediator) {}
  
  void runTrials() {
    cout << "Clinical Trials: Running clinical trials..." << endl;
    mediator->notify("ClinicalTrials", "Clinical Trial Completed");
  }
  
  void completeTask() override {}
};

// Concrete Department: Quality Assurance
class QualityAssurance : public Department {
public:
  QualityAssurance(Mediator* mediator) : Department(mediator) {}
  
  void checkQuality() {
    cout << "Quality Assurance: Checking product quality..." << endl;
    mediator->notify("QualityAssurance", "Quality Check Completed");
  }
  
  void completeTask() override {}
};

// Concrete Department: Regulatory Affairs
class RegulatoryAffairs : public Department {
public:
  RegulatoryAffairs(Mediator* mediator) : Department(mediator) {}
  
  void prepareDocuments() {
    cout << "Regulatory Affairs: Preparing regulatory documents..." << endl;
    mediator->notify("RegulatoryAffairs", "Documents Prepared");
  }
  
  void completeTask() override {}
};

// Concrete Department: Legal Department
class LegalDepartment : public Department {
public:
  LegalDepartment(Mediator* mediator) : Department(mediator) {}
  
  void reviewDocuments() {
    cout << "Legal Department: Reviewing regulatory documents..." << endl;
    mediator->notify("LegalDepartment", "Legal Review Completed");
  }
  
  void completeTask() override {}
};

// Concrete Mediator Method Implementation
void RegulatoryMediator::notify(const string& sender, const string& event) {
  if (event == "Clinical Trial Completed") {
    cout << "Mediator: Clinical Trial completed. Notifying Quality Assurance." << endl;
    if (quality_assurance) {
      quality_assurance->checkQuality();
    }
  } else if (event == "Quality Check Completed") {
    cout << "Mediator: Quality Check completed. Notifying Regulatory Affairs." << endl;
    if (regulatory_affairs) {
      regulatory_affairs->prepareDocuments();
    }
  } else if (event == "Documents Prepared") {
    cout << "Mediator: Documents prepared. Notifying Legal Department." << endl;
    if (legal) {
      legal->reviewDocuments();
    }
  } else if (event == "Legal Review Completed") {
    cout << "Mediator: Legal Review completed. Drug is ready for final approval." << endl;
  }
}

int main() {
  // Create Mediator
  auto mediator = make_shared<RegulatoryMediator>();
  
  // Create Departments
  auto clinical_trials = make_shared<ClinicalTrials>(mediator.get());
  auto quality_assurance = make_shared<QualityAssurance>(mediator.get());
  auto regulatory_affairs = make_shared<RegulatoryAffairs>(mediator.get());
  auto legal = make_shared<LegalDepartment>(mediator.get());
  
  // Set Departments in the Mediator
  mediator->setClinicalTrials(clinical_trials.get());
  mediator->setQualityAssurance(quality_assurance.get());
  mediator->setRegulatoryAffairs(regulatory_affairs.get());
  mediator->setLegalDepartment(legal.get());
  
  // Run process
  clinical_trials->runTrials();
  
  return 0;
}
